firedrake/interpolation.py

```diff
@@ -254,20 +293,48 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
                 raise NotImplementedError("Can only interpolate Python kernels with Lagrange elements")
             pts, = dual.pt_dict.keys()
             to_pts.append(pts)
-        kernel, oriented, needs_cell_sizes, coefficients = compile_python_kernel(expr, to_pts, to_element, V, coords)
+        kernel, oriented, needs_cell_sizes, coefficients = compile_python_kernel(expr, to_pts, to_element, V, source_coords_coeff)
         first_coeff_fake_coords = False
     else:
         raise RuntimeError("Attempting to evaluate an Expression which has no value.")
 
-    cell_set = coords.cell_set
+    cell_set = target_mesh.coordinates.cell_set
     if subset is not None:
         assert subset.superset == cell_set
         cell_set = subset
     parloop_args = [kernel, cell_set]
 
     if first_coeff_fake_coords:
-        # Replace with real coords coefficient
-        coefficients[0] = coords
+        # Replace with real source_coords_coeff
+        coefficients[0] = source_coords_coeff
+
+    if trans_mesh:
+        # NOTE: TSFC will sometimes drop run-time arguments in generated
+        # kernels if they are deemed not-necessary.
+        # FIXME: Checking for argument name in the inner kernel to decide
+        # whether to add an extra coefficient is a stopgap until
+        # compile_expression_dual_evaluation and compile_python_kernel
+        #   (a) output a coefficient map to indicate argument ordering in
+        #       parloops as `compile_form` does and
+        #   (b) allow the dual evaluation related coefficients to be suplied to
+        #       them rather than having to be added post-hoc (likely by
+        #       replacing `to_element` with a CoFunction/CoArgument as the
+        #       target `dual` which would contain `dual` related
+        #       coefficient(s))
+        if rt_var_name in [arg.name for arg in kernel.code[name].args]:
+            # Add the coordinates of the target mesh quadrature points in the
+            # source mesh's reference cell as an extra argument for the inner
+            # loop. (With a vertex only mesh this is a single point for each
+            # vertex cell.)
+            coefficients = coefficients + [target_mesh.reference_coordinates]
+        # In the trans-mesh case we loop over the target mesh cells so we need
+        # to compose a map that takes us from their cells to the source mesh
+        # cells, and (where necessary) to the function space nodes on the
+        # source mesh.
+        if isinstance(target_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology):
+            target_mesh_source_mesh_map = target_mesh.cell_parent_cell_map
+        else:
+            raise NotImplementedError("Can only interpolate onto a Vertex Only Mesh")
```

Comment:
Maybe we should include this check when we define `trans_mesh`.

Reply:
Agreed